### PR TITLE
run: prompt when removing outputs

### LINKED ISSUE #2027: run: warn and/or prompt user when deleting an output
The following command deletes data/clean -- it would be nice if there was a warning or something since there may already be data in data/clean.  The solution was to modify the script so that the data/clean directory was created by the script. 

dvc run -d data/raw/ -d src/featherize_data.py --outs data/clean/ python src/featherize_data.py

<<Hi @david8381 ! That happens because dvc run is trying to ensure that your command is the one creating your output.
So that when you run dvc repro later, it will be able to fully reproduce the output.
So you need to make your script create that directory.>>

### PR DESCRIPTION
**Pull request by mroutis**
Close #2027

* [x] Have you followed the guidelines in our
      [Contributing document](https://dvc.org/doc/user-guide/contributing)?

* [ ] Does your PR affect documented changes or does it add new functionality
      that should be documented? If yes, have you created a PR for
      [dvc.org](https://github.com/iterative/dvc.org) documenting it or at
      least opened an issue for it? If so, please add a link to it.

-----

[asciicast](https://asciinema.org/a/253454)

Uses `--remove-outs` to prevent prompting.

### CODE REVIEW

File: dvc/repo/run.py [hunk 1]
```diff
@@ -66,6 +70,35 @@ def run(
 
     with self.state:
         if not no_exec:
+            if (stage.cmd or stage.is_import) and not stage.locked:
+                existing = [
```
Comment (shcheklein) [inline dvc/repo/run.py]: looks like this list is calculated in a different way than in `stage.remove_outs`

Comment (shcheklein) [reply dvc/repo/run.py]: we also calculate this list no matter what, even if `--remove-outs` is specified. I'm not sure how expensive the changed_cache could be. Or can become in the future.

File: dvc/repo/run.py [hunk 2]
```diff
@@ -66,6 +70,35 @@ def run(
...
+                    out
+                    for out in stage.outs
+                    if os.path.exists(str(out))
+                    if out.changed_cache()
+                ]
+
+                statement = (
```
Comment (shcheklein) [inline dvc/repo/run.py]: do we show it even if `existing` is empty?

File: dvc/repo/run.py [hunk 3]
```diff
@@ -66,6 +70,35 @@ def run(
...
+                    existing
+                    and not remove_outs
+                    and not prompt.confirm(statement)
+                ):
+                    raise DvcException(
+                        "cannot destroy without a confirmation from the user."
+                        " Use --remove-outs if you don't want to be asked."
+                    )
+
+                stage.remove_outs(ignore_remove=False, force=False)
```
Comment (shcheklein) [inline dvc/repo/run.py]: I think we can move the logic inside remove_outs, add a flag and centralize the logic that calculates the list of files to remove

File: dvc/repo/run.py (hunk 1 above)
Comment (shcheklein) [reply dvc/repo/run.py]: Was the intention to show only outputs that are not in cache? Then the message is confusing. Because it actually should show _all_ outputs (not persisted) the way it's written.

File: dvc/repo/run.py (hunk 2 above)
Comment (shcheklein) [reply dvc/repo/run.py]: should we actually check if out exists on remote?

File: dvc/repo/imp_url.py [hunk 4]
```diff
@@ -27,6 +27,7 @@ def imp_url(self, url, out=None, resume=False, fname=None, erepo=None):
     self.check_dag(self.stages() + [stage])
 
     with self.state:
+        stage.remove_outs(ignore_remove=False, force=False)
```
Comment (shcheklein) [inline dvc/repo/imp_url.py]: it should probably behave the same way as a regular run. Ask for confirmation? Especially if data is not cached or saved to remote at least

**Comment by shcheklein** [comment]
should we have an option for `repro` to pass down to the stages to skip the confirmation?

**Comment by shcheklein** [comment]
create a ticket for the documentation with the description of changes or update the docs please :)

File: dvc/repo/run.py (hunk 2 above)
Comment (shcheklein) [reply dvc/repo/run.py]: will it ask about --outs-no-cache as well? usually files that are put under git control?

File: dvc/repo/run.py [hunk 5]
```diff
@@ -66,6 +70,35 @@ def run(
 
     with self.state:
         if not no_exec:
+            if (stage.cmd or stage.is_import) and not stage.locked:
+                existing = [
+                    out
+                    for out in stage.outs
+                    if os.path.exists(str(out))
```
Comment (pared) [inline dvc/repo/run.py]:
```suggestion
                    if self.tree.exists(out.path_info)
```

File: dvc/repo/run.py (hunk 1 above)
Comment (mroutis) [reply dvc/repo/run.py]: Agree, @shcheklein , message is misleading, the intention was to show it only when the output already exists and is not cached, but the latter doesn't make a lot of sense (in what cases would the output be cached? just when the output is already on another DVC-file).

I'll remove the `out.changed_cache` condition.

File: dvc/repo/run.py (hunk 2 above)
Comment (mroutis) [reply dvc/repo/run.py]: > do we show it even if existing is empty?

No, we check this before prompting the user

> should we actually check if out exists on remote?

:thinking: Yes! Because `stage.remove_outs` goes to each remote and deletes the output.

> will it ask about --outs-no-cache as well? usually files that are put under git control?

Yes, it will, because the point of this message is to warn the user about DVC removing files, even if you `dvc run --outs-no-cache` it will remove the outputs to make sure that the command is the one creating the output.

File: dvc/repo/run.py (hunk 3 above)
Comment (mroutis) [reply dvc/repo/run.py]: @shcheklein , we have a `safe_remove` method already, but it handles a single path and the exception is suggesting the use of `--force` instead of `--remove-outs`, that's why I moved the logic outside. Also, the `force` argument in `remove_outs` is used for removing _persisted outputs_, maybe adding a `prompt=False` argument? I'll try different stuff

File: dvc/repo/imp_url.py (hunk 4 above)
Comment (mroutis) [reply dvc/repo/imp_url.py]: Ok

**Comment by mroutis** [comment]
> should we have an option for `repro` to pass down to the stages to skip the confirmation?

@shcheklein , this doesn't affect `repro`, it is `run` specific

File: dvc/repo/run.py (hunk 2 above)
Comment (shcheklein) [reply dvc/repo/run.py]: so, if I have a, let's say, a metric file and run dvc repro/dvc run - it will be asking me every time ... not sure this is the best UX ...  I don't know if there other options though

File: dvc/stage.py [hunk 6]
```diff
@@ -493,18 +496,7 @@ def create(
         stage.wdir = wdir
         stage.path = path
 
-        # NOTE: remove outs before we check build cache
```
Comment (efiop) [inline dvc/stage.py]: Why did we move this outside?

Comment (efiop) [reply dvc/stage.py]: The change you are working on should work for `import-url` too.

Comment (efiop) [reply dvc/stage.py]: ?

Comment (mroutis) [reply dvc/stage.py]: > Why did we move this outside all the way up to dvc/repo/run.py?

I'm deleting this because it implies an old / deprecated use of `--remove-outs`.

> The change you are working on should work for import-url too.

Yep, will consider it in the next commit :+1:

File: dvc/repo/run.py (hunk 2 above)
Comment (mroutis) [reply dvc/repo/run.py]: :sweat_smile: this is what I'm facing right now, all the tests related to metrics are failing because of this assumption, @shcheklein , I'll think about it

Comment (efiop) [reply dvc/repo/run.py]: It is using `safe_remove` so most of the time for cached outputs(metrics included) it should be fine. For not cached ones, one might not want to lose it silently, so it makes sense to ask about it there as well. As to repro, I don't think this PR is changing anything for it at all, nor should it.

Btw, the first case when output is already cached, the referenced code doesn't make sense, because it checks only for _existance_, but if the output is already cached, we should not ask about deleting it. The related discussion was about user losing the old dvc file, for which we've reached the conclusion that we should not delete the dvc file in create() right away, but rather just overwrite it when and if the command is successful(which could probably be achieved by simply removing os.unlink() in Stage.create()).

File: dvc/output/base.py [hunk 7]
```diff
@@ -297,8 +297,9 @@ def checkout(self, force=False, progress_callback=None, tag=None):
             progress_callback=progress_callback,
         )
 
-    def remove(self, ignore_remove=False):
-        self.remote.remove(self.path_info)
+    def remove(self, ignore_remove=False, ask=False):
```
Comment (efiop) [inline dvc/output/base.py]: we can actually call it `force` here, don't we?

Comment (mroutis) [reply dvc/output/base.py]: okok :+1:

File: dvc/output/base.py [hunk 8]
```diff
@@ -297,8 +297,9 @@ def checkout(self, force=False, progress_callback=None, tag=None):
             progress_callback=progress_callback,
         )
 
-    def remove(self, ignore_remove=False):
-        self.remote.remove(self.path_info)
+    def remove(self, ignore_remove=False, force=True):
+        self.cache.safe_remove(self.path_info, force=force)
```
Comment (efiop) [inline dvc/output/base.py]: And what if output doesn't use cache? I'm pretty sure you've mentioned this :)

Comment (mroutis) [reply dvc/output/base.py]: I did, but removed it since the tests passed :sweat_smile: 
I still think it make sense, I'll add it.

File: dvc/output/base.py [hunk 9]
```diff
@@ -297,8 +297,12 @@ def checkout(self, force=False, progress_callback=None, tag=None):
             progress_callback=progress_callback,
         )
 
-    def remove(self, ignore_remove=False):
-        self.remote.remove(self.path_info)
+    def remove(self, ignore_remove=False, force=True):
+        if self.cache:
+            self.cache.safe_remove(self.path_info, force=force)
+        else:
+            self.remote.remove(self.path_info)
```
Comment (efiop) [inline dvc/output/base.py]: If file doesn't use cache, then we should also confirm if we want to delete it, right?

Comment (mroutis) [reply dvc/output/base.py]: no, wait, @efiop , if a remote doesn't use `cache` it can't use `external-outputs`, therefore, there's no chance of removing stuff

Comment (mroutis) [reply dvc/output/base.py]: does it make sense to remove this conditional and stay with `self.cache.safe_remove`?

Comment (efiop) [reply dvc/output/base.py]: Not sure how external outputs are related here. If output doesn't use cache, we are still removing it currently, right? If so, we need to ask user about it, right?

Comment (mroutis) [reply dvc/output/base.py]: Yes, but not _using_ cache is not the same as not _having_ cache.
If you specify `dvc run --outs-no-cache foo` it will _have_ a cache but it won't use it.

Comment (mroutis) [reply dvc/output/base.py]: Discussed privately that `dvc run --outs-no-cache` with external outputs that don't have cache configured will raise this issue.

File: dvc/command/imp_url.py [hunk 10]
```diff
@@ -57,4 +60,10 @@ def add_parser(subparsers, parent_parser):
     import_parser.add_argument(
         "-f", "--file", help="Specify name of the DVC-file it generates."
     )
+    import_parser.add_argument(
+        "--remove-outs",
+        action="store_true",
+        default=False,
+        help="Confirm removal of existing files used as outputs",
```
Comment (shcheklein) [inline dvc/command/imp_url.py]: took me a while to understand the semantics. Let's do something like "Always remove existing outputs w/o asking for a confirmation data is not in cache and can be lost"

File: dvc/command/imp_url.py [hunk 11]
```diff
@@ -57,4 +60,10 @@ def add_parser(subparsers, parent_parser):
     import_parser.add_argument(
         "-f", "--file", help="Specify name of the DVC-file it generates."
     )
+    import_parser.add_argument(
+        "--remove-outs",
```
Comment (shcheklein) [inline dvc/command/imp_url.py]: not sure this us the best name for the option. It feels like w/o this option we are not removing outputs at all.

File: dvc/command/imp_url.py [hunk 12]
```diff
@@ -14,7 +14,10 @@ class CmdImportUrl(CmdBase):
     def run(self):
         try:
             self.repo.imp_url(
-                self.args.url, out=self.args.out, fname=self.args.file
```
Comment (shcheklein) [inline dvc/command/imp_url.py]: if we change `imp_url` we should probably change `import` as well?

File: dvc/remote/base.py [hunk 13]
```diff
@@ -630,6 +630,9 @@ def exists_with_progress(chunks):
         return list(set(checksums) & set(self.all()))
 
     def already_cached(self, path_info):
+        if not self.cache:
```
Comment (shcheklein) [inline dvc/remote/base.py]: why this change? was there a bug before?

File: scripts/completion/dvc.zsh [hunk 14]
```diff
@@ -235,7 +235,7 @@ _dvc_run=(
   {-y,--yes}"[Deprecated, use --overwrite-dvcfile instead]"
   "--overwrite-dvcfile[Overwrite existing DVC-file without asking for confirmation.]"
   "--ignore-build-cache[Run this stage even if it has been already ran with the same command/dependencies/outputs/etc before.]"
-  "--remove-outs[Deprecated, this is now the default behavior]"
+  "--remove-outs[Confirm removal of existing files used as outputs]"
```
Comment (shcheklein) [inline scripts/completion/dvc.zsh]: update bash prompt as well, create a ticket for docs

File: dvc/remote/base.py (hunk 13 above)
Comment (mroutis) [reply dvc/remote/base.py]: `output.remote` doesn't have a `path_info` and crashes the whole thing, the alternative was to use `output.cache.already_cached` (as `cache` has `path_info`), however, when a remote is not configured (for example, non-cached external outputs) `cache` would be None.

Remembering these edge cases is error prone, so now you can use `output.remote.already_cached` and you'll be covered.

File: dvc/command/imp_url.py (hunk 11 above)
Comment (mroutis) [reply dvc/command/imp_url.py]: true, @shcheklein , what about `--force` ? maybe `--no-ask-remove`? any suggestions?

File: dvc/command/imp_url.py (hunk 12 above)
Comment (mroutis) [reply dvc/command/imp_url.py]: make sense :+1:

File: dvc/command/imp_url.py (hunk 11 above)
Comment (efiop) [reply dvc/command/imp_url.py]: I'm actually fine with reusing this option, since we already have it. But we could also use `--force-remove-outs` instead.

Comment (mroutis) [reply dvc/command/imp_url.py]: @shcheklein , does `--force-remove-outs` make sense to you?

Comment (shcheklein) [reply dvc/command/imp_url.py]: yep, I like it. I'm not sure I understand the compatibility with `dvc run` point

File: dvc/exceptions.py [hunk 15]
```diff
@@ -178,8 +178,9 @@ def __init__(self, stages):
 class ConfirmRemoveError(DvcException):
     def __init__(self, path):
         super(ConfirmRemoveError, self).__init__(
-            "unable to remove '{}' without a confirmation from the user. Use "
-            "'-f' to force.".format(path)
+            "unable to remove '{}' without a confirmation from the user."
+            " Use '--force' (repro) or '--remove-outs' (run) to confirm"
```
Comment (shcheklein) [inline dvc/exceptions.py]: don't forget to update the name here

File: dvc/remote/base.py (hunk 13 above)
Comment (shcheklein) [reply dvc/remote/base.py]: so, what is the difference with was here before? I'm not sure I understand this still.

File: dvc/remote/local/__init__.py [hunk 16]
```diff
@@ -195,7 +195,7 @@ def already_cached(self, path_info):
         if not current_md5:
             return False
 
-        return not self.changed_cache(current_md5)
+        return not self.cache.changed_cache(current_md5)
```
Comment (shcheklein) [inline dvc/remote/local/__init__.py]: again, what's the difference here. All this relationship between remote, cache, and output is very confusing tbh.

File: dvc/stage.py [hunk 17]
```diff
@@ -487,41 +509,19 @@ def create(repo, **kwargs):
...
-                " It is now the default behavior,"
-                " so there's no need to use this option anymore."
-            )
-            stage.remove_outs(ignore_remove=False)
-            logger.warning("Build cache is ignored when using --remove-outs.")
-            ignore_build_cache = True
-        else:
-            stage.unprotect_outs()
-
-        if os.path.exists(path) and any(out.persist for out in stage.outs):
```
Comment (shcheklein) [inline dvc/stage.py]: is it safe to remove?

File: dvc/stage.py [hunk 18]
```diff
@@ -299,7 +307,11 @@ def remove_outs(self, ignore_remove=False, force=False):
                         out=out, stage=self.relpath
                     )
                 )
-                out.remove(ignore_remove=ignore_remove)
+
+                if force:
+                    ask = False
+
+                out.remove(ignore_remove=ignore_remove, force=not ask)
```
Comment (shcheklein) [inline dvc/stage.py]: all this naming is super confusing - force, aks, confirm mean different things even in the same function sometimes

Comment (mroutis) [reply dvc/stage.py]: yep, can't agree more, @shcheklein , but changing this will require a big refactor, `force` means different things, for example: `ask` / `prompt` / `confirm` are different names for the same thing, the problem is that `prompt` is the module, `confirm` is the method and `_ask` is the implementation beneath `confirm` :sweat_smile:

We can address this in another issue.

File: dvc/remote/base.py (hunk 13 above)
Comment (mroutis) [reply dvc/remote/base.py]: We are calling `stage.remove_outs` on `Stage.create` (we are deleting the outputs to make sure the command is generating them)

```python
        # dvc/stage.py (create)

        # Removing outputs for reproducibility sake,
        # it is expected that the `cmd` will create them.
        if stage.cmd:
            confirm = kwargs.get("remove_outs", False)
            stage.remove_outs(ask=not confirm)
```

I added the `ask` parameter to `remove_outs` (previously, it just removed the outputs without asking (calling `output.remove`), but now, `output.remove` calls `output.safe_remove` if `ask` is set to `true` (this method asks the user for confirmation before it deletes the output).

```python
    # dvc/stage.py
    def remove_outs(self, ignore_remove=False, force=False, ask=False):
        """Used mainly for `dvc remove --outs` and :func:`Stage.reproduce`."""
        for out in self.outs:
            if out.persist and not force:
                out.unprotect()
            else:
                logger.debug(
                    "Removing output '{out}' of '{stage}'.".format(
                        out=out, stage=self.relpath
                    )
                )

                if force:
                    ask = False

                out.remove(ignore_remove=ignore_remove, force=not ask)
```

```python
    # dvc/output/base.py
    def remove(self, ignore_remove=False, force=True):
        self.remote.safe_remove(self.path_info, force=force)
```

This is were things start to get confusing with the `ask`, `force`, and `confirm`.

* For `stage.remove_outs`:
  * `force` means to remove persisting outputs.
  * `ask` means to _ask for confirmation_ when removing outputs (by default set to false (to not affect the previously defined interface)

For `output.remove`:
  * `force` is passed directly to `safe_remove`, and it means to not ask for confirmation while "safe removing".

Now, `output.remove` calls `remote.safe_remove`.
For `safe_remove` to work, it needs a `cache_dir` (because safe remove checks if the output is on the cache, if it has no cache it crashes -- this was not a bug before because we were not using safe remove on `stage.create` but now, if the remote has no `cache_dir` it will break, that's why I'm checking if the cache exists and sending methods to the `cache` object instead of the `remote` one).

Imagine the following case:

```bash
dvc remote add ssh ssh://localhost
dvc run --outs-no-cache remote://ssh/file.txt
```

The stage file is going to have one output.
This output doesn't require any cache.

This is where the `if not self.cache` is relevant.

I hope this helps, it took me a while to get everything on my head.

File: dvc/remote/local/__init__.py (hunk 16 above)
Comment (mroutis) [reply dvc/remote/local/__init__.py]: @shcheklein , it is confusing, indeed. This solve the same problem explained above.

Comment (efiop) [reply dvc/remote/local/__init__.py]: @mroutis So why not `self.changed_cache`? It should do the same thing, no? Why do we need to access self.cache.changed_cache directly?

File: dvc/exceptions.py [hunk 19]
```diff
@@ -178,8 +178,9 @@ def __init__(self, stages):
 class ConfirmRemoveError(DvcException):
     def __init__(self, path):
         super(ConfirmRemoveError, self).__init__(
-            "unable to remove '{}' without a confirmation from the user. Use "
-            "'-f' to force.".format(path)
+            "unable to remove '{}' without a confirmation from the user."
+            " Use '--force' (repro) or '--force-remove-outs' (run) to confirm"
```
Comment (efiop) [inline dvc/exceptions.py]: Do we need to check that on `repro` though? Didn't we agree that it is allowed for `repro` to do that silently? `--force` for `repro` has a totally different meaning than `--force-remove-outs`.

File: dvc/output/base.py [hunk 20]
```diff
@@ -297,8 +297,9 @@ def checkout(self, force=False, progress_callback=None, tag=None):
             progress_callback=progress_callback,
         )
 
-    def remove(self, ignore_remove=False):
-        self.remote.remove(self.path_info)
+    def remove(self, ignore_remove=False, force=True):
+        self.remote.safe_remove(self.path_info, force=force)
```
Comment (efiop) [inline dvc/output/base.py]: But `self.remote` might not be the same as `self.cache` and `self.cache` is the one that actually has cache and can check if cache for that path_info is present. I'm pretty sure you've had a `if self.use_cache` here previously, no?

File: dvc/remote/local/__init__.py (hunk 16 above)
Comment (mroutis) [reply dvc/remote/local/__init__.py]: `self` is a remote with no `path_info`, due to the way outputs are instantiated:
```python
        self.remote = remote or self.REMOTE(self.repo, {})
```

If we call `changed_cache` directly it will fail on this one:
```python
    def checksum_to_path_info(self, checksum):
        return self.path_info / checksum[0:2] / checksum[2:]
```

Comment (efiop) [reply dvc/remote/local/__init__.py]: @mroutis I guess you meant that
```
        return not self.cache.changed_cache(current_md5)
```
is different from
```
    def changed_cache(self):                           
        if not self.use_cache or not self.checksum:    
            return True                                
                                                       
        return self.cache.changed_cache(self.checksum) 
```
because in the self.changed_cache we are using self.checksum wich might not match current_md5?

File: dvc/output/base.py (hunk 20 above)
Comment (mroutis) [reply dvc/output/base.py]: This is solved on this one:
```diff
    def already_cached(self, path_info):
        assert path_info.scheme in ["", "local"]

        current_md5 = self.get_checksum(path_info)

        if not current_md5:
            return False

-        return not self.changed_cache(current_md5)
+        return not self.cache.changed_cache(current_md5)
```

It doesn't matter if you call it with `remote` or `cache` (It thought this was a better interface / less error prone) :sweat_smile:  at the end, it will call the `cache`, the tricky part is to get that you can call `cache.cache.cache....` multiple times and it will return the same instance

File: dvc/remote/local/__init__.py (hunk 16 above)
Comment (mroutis) [reply dvc/remote/local/__init__.py]: @efiop , where did you get the `self.use_cache` and `self.checksum`? I'm not sure if we are talking about the same `changed_cache` :sweat_smile: there's one in `output/base.py` and another one in `remote/base.py`. I'm talking about latter one. (this change is on `remote/base.py` -- just clearing it out of the way, it might result confusing)

File: dvc/output/base.py (hunk 20 above)
Comment (mroutis) [reply dvc/output/base.py]: The `if self.use_cache` is on `changed_cache` already :thinking:

File: dvc/exceptions.py (hunk 19 above)
Comment (mroutis) [reply dvc/exceptions.py]: :thinking: that's true, @efiop , when `--force` is used with `repro` it reproduces the stage even if the stage didn't _change_.

I'm not sure why the error message was like that.

Comment (mroutis) [reply dvc/exceptions.py]: It might be a bug:
```python
    def run(self, dry=False, no_commit=False, force=False):
        if (self.cmd or self.is_import) and not self.locked and not dry:
            self.remove_outs(ignore_remove=False, force=False)
```

I think it should have been `self.remove_outs(ignore_remove=False, force=force)`

File: dvc/remote/local/__init__.py (hunk 16 above)
Comment (efiop) [reply dvc/remote/local/__init__.py]: @mroutis Aah, indeed, i was looking at `output` for some reason. Sorry! So looks like we've had a bug here before, right?

File: dvc/stage.py (hunk 18 above)
Comment (efiop) [reply dvc/stage.py]: @mroutis Seems like `force` has the meaning of something like `remove_persistent`? Maybe let's switch them up? Your `ask` -> `force` and `force` -> `remove_persistent`. Would that make sense?

Comment (mroutis) [reply dvc/stage.py]: @efiop , lol, it make sense, I'll do that ^
